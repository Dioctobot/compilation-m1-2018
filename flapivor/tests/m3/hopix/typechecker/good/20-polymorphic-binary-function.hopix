def apply : forall a, b. (a -> b) * a -> b (f, x) = f (x)

def curry : forall a, b, c. (a -> b -> c) -> (a * b -> c) (f) =
    def fc : a * b -> c (x, y) = (f (x)) (y);
    fc

def uncurry : forall a, b, c. (a * b -> c) -> (a -> b -> c) (f) =
    def fu : a -> b -> c (x) =
      (def g : b -> c (y) = f (x, y);
      g);
    fu

def id_int : Int -> Int (x) = x

val main : Int =
   (curry<Int -> Int, Int, Int> (uncurry<Int -> Int, Int, Int> (apply<Int, Int>))) (id_int, 37)