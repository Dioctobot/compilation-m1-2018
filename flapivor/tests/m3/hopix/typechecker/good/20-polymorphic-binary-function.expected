apply : forall a, b. (a -> b) * a -> b
curry : forall a, b, c. (a -> (b -> c)) -> (a * b -> c)
uncurry : forall a, b, c. (a * b -> c) -> (a -> (b -> c))
id_int : Int -> Int
main : Int
